modules/core/src/copy.cpp

```diff
@@ -319,10 +319,10 @@ void Mat::copyTo( OutputArray _dst ) const
             Size sz = getContinuousSize(*this, dst);
             size_t len = sz.width*elemSize();
 
-#if defined HAVE_IPP
+#if defined HAVE_IPP && !(defined __LP64__ || defined _WIN64)
             CV_IPP_CHECK()
             {
-                if (ippiCopy_8u_C1R(sptr, (int)step, dptr, (int)dst.step, ippiSize((int)len, sz.height)) >= 0)
+                if (ippiCopy_8u_C1R(sptr, (int)step, dptr, (int)dst.step, ippiSize(len, sz.height)) >= 0)
```

Comment:
@alalek, the scenario is as following:
1. After several calls to `push_back` (each adds ~1,7Gb), the `reserve` method is called.
2. `reserve` allocates new Mat and tries to copy data from old Mat (with `copyTo`), data size is bigger than 4Gb  and some data is not copied

I looked over the code for the `ippiSize` structure initialization with similar truncation and had not found. I'll try the suggested 4Gb boundary check tomorrow.